feat: 구독 요청 API

soo96 opened the pull request:
## 요약

- 구독 요청 API

Close: #32 

## 변경 사항

- [x]  로그인된 사용자인지 확인. (JWT 토큰 검증)
    - [x]  로그인된 사용자가 아니라면 요청 거부.
    - [x]  “인증되지 않은 사용자입니다.” 에러메시지와 함께 **`401 Unauthorization`** 에러 응답
    - [x]  JWT 토큰에서 `userId` 정보 확인
- [x]  등록 요청한 **채널 URL** 주소가 올바른지 확인.
    - [x]  각 플랫폼에 요청을 보내서 유효한 URL 인지 확인.
    - [x]  URL주소가 올바르지 않다면 요청 거부.
        - [x]  “잘못된 채널URL입니다.” 에러메시지와 함께 **`400 Bad Request`** 에러 응답
- [x]  Streamer, Subscription 테이블에 정보 등록
    - [x]  정상 등록되었다면 성공 응답을 전송한다.
        - [x]  **`201 Created`** 성공 응답
    - [x]  이미 구독중인 채널URL 정보이면 실패 응답을 전송한다.
        - [x]  “이미 구독중인 채널입니다.” 에러메시지와 함께 **`409 Conflict`** 에러 응답
    - [x]  최대 구독 채널수를 초과한 경우
        - [x]  “더 이상 채널을 구독할 수 없습니다.” 에러메시지와 함께 **`403 Forbiden`** 에러 응답
    - [x]  그외 기타 오류들은 실패 응답을 전송한다.
        - [x]  **`500 Internal Server Error`** 에러 응답

## 체크리스트

- [ ] 이해하기 어려운 코드에는 주석을 추가했습니다  
- [ ] 관련 문서를 수정했습니다  
- [ ] 새로운 경고가 발생하지 않습니다  
- [ ] 수정한 기능 또는 추가한 기능에 대해 테스트를 작성했습니다  
- [ ] 의존성 있는 변경 사항이 병합 및 배포되었습니다

## 관련 이슈

- #32 

## 리뷰 반영 사항

<!-- PR리뷰 반형 사항 -->

<details>
<summary><strong>선택 항목 펼치기</strong></summary>

## 스크린샷

<!-- 시각적인 변경 사항이 있다면 스크린샷이나 GIF를 포함해 주세요. 리뷰어가 쉽게 이해할 수 있습니다. -->

## 테스트 방법

<!-- PR에서 수행한 변경 사항을 어떻게 테스트할 수 있는지 설명해 주세요. 리뷰어가 직접 확인할 수 있도록 돕습니다. -->

## 리뷰어 참고사항

<!-- 리뷰어에게 특별히 알려야 할 사항이나 고려할 점이 있다면 여기에 작성해 주세요. -->

</details>

soo96 committed:
refactor: JwtPayload 타입 정의 및 구조 변경 (email 제거, sub 해제)

soo96 committed:
feat: UserId 데코레이터 생성

soo96 committed:
refactor: Prisma 스키마 변경

- Stremaer 테이블 channel_id 유니크 추가
- Subscription 테이블 is_connected not null 추가

soo96 committed:
feat: CHZZK, YOUTUBE 채널정보 가져오는 기능 구현

soo96 committed:
feat: 채널 구독 요청 기능 구현

soo96 committed:
test: 구독 테스트 모듈에 STREAMER_REPOSITORY, STREAMING_SERVER_CLIENT provider 추가

soo96 committed:
fix: 구독 목록 조회 시 user_id 및 is_connected=true 조건 추가

soo96 committed:
feat: 최대 구독수 제한 로직 추가

soo96 committed:
test: 테스트 코드 getSubscriptions에 userId 적용

soo96 committed:
chore: dev브랜치 최신사항 반영

soo96 committed:
feat: 구독 상태 변경 시 Redis 플래그 true로 설정

soo96 committed:
test: SubscriptionService 테스트코드 CACHE_SERVICE 의존성 주입 추가

smileboy0014 commented on src/domain/subscription/subscription.repository.ts:
```diff
@@ -1,7 +1,13 @@
 
 export interface SubscriptionRepository {
-  getSubscriptions(): Promise<SubscriptionWithChannelResult[]>;
+  getSubscriptions(userId: number): Promise<SubscriptionWithChannelResult[]>;
+  findSubscription(userId: number, streamerId: number): Promise<SubscriptionEntity | null>;
```
```suggestion
  getSubscriptions(userId: number): Promise<SubscriptionWithChannelResult[]>;
  getSubscription(userId: number, streamerId: number): Promise<SubscriptionEntity | null>;
```
위에서는 `getSubscriptions` 표현을 쓰셨는데, 밑에서는 `findSubscription` 표현을 쓰셔서 조금 어색한 것 같습니다.
`getSubscription` 을 쓰는게 변수명이 더 낫지 않을까요?

soo96 replied:
조회 메서드는 `get` 접두사를 사용하도록 수정했습니다! 
감사합니다.🙂

smileboy0014 commented on src/domain/subscription/subscription.service.ts:
```diff
@@ -1,15 +1,76 @@
+        streamer.streamerId
+      );
+
+      if (subscription && subscription.isConnected) {
+        throw new DomainCustomException(
+          HttpStatus.CONFLICT,
+          DomainErrorCode.DUPLICATED_SUBSCRIPTION
+        );
+      }
```
```suggestion
subscription.isConnected();
```
절차지향적인 코드 느낌이 많이 납니다.
도메인 중심적인 코드로 한번 리팩토링 해보면 어떨까요??

```javascript
export class SubscriptionEntity {
  constructor(
    public readonly subscriptionId: number,
    public readonly userId: number,
    public readonly streamerId: number,
    public readonly isConnected: boolean,
    public readonly createdAt: Date,
    public readonly updatedAt: Date
  ) {}

  isConnected(): void {
    if (this && this.isConnected) {
        throw new DomainCustomException(
          HttpStatus.CONFLICT,
          DomainErrorCode.DUPLICATED_SUBSCRIPTION
        );
      } 
  }
}
```

soo96 replied:
- SubscriptionEntity
```ts
connect() {
    if (this.isConnected) {
      throw new DomainCustomException(HttpStatus.CONFLICT, DomainErrorCode.DUPLICATED_SUBSCRIPTION);
    }

    this.isConnected = true;

    return this;
  }
```

이런식으로 적용해봤습니다. 🙂

smileboy0014 commented on src/infrastructure/subscription/subscription.core-repository.ts:
```diff
@@ -2,13 +2,84 @@ import { Injectable } from '@nestjs/common';
+    });
+  }
+
+  async createSubscriptionWithStreamer(userId: number, channelInfo: ChannelInfo): Promise<void> {
+    const { channelId, channelName, platform } = channelInfo;
+
+    await this.prisma.$transaction(async (tx) => {
+      const streamer = await tx.streamer.create({
+        data: {
+          channel_id: channelId,
+          channel_name: channelName,
+          platform,
+        },
+      });
+
+      await tx.subscription.create({
+        data: {
+          user_id: userId,
+          streamer_id: streamer.streamer_id,
+          is_connected: true,
+        },
+      });
+    });
+  }
```
streamer와 subscription은 연결되어 있는 도메인 관계가 아닌데, 여기서 트랜잭션을 묶어서 처리하는게 맞을까요??
streamerRepository 와 subscriptionRepository에 각각 메서드를 정의해서 application 계층에서 조합해서 트랜잭션을 묶어주는게 맞지 않을까요?

soo96 replied:
@smileboy0014 
리뷰 감사드립니다.🙏🏻
말씀해주신 것처럼 `streamer`와 `subscription`은 도메인 상 서로를 포함하거나 종속하지 않는 동등한 수준의 별개 도메인이기 때문에, 각 도메인 레이어(`StreamerRepository`, `SubscriptionRepository`)에는 자기 영역의 독립적인 행위만 정의하는 것이 적절하다고 판단했습니다.

따라서 두 도메인을 함께 다뤄야 하는 흐름은 `애플리케이션 계층`의 조합 책임이라고 보고, 해당 로직을 `UseCase`로 분리하여 트랜잭션 범위 안에서 처리하도록 리팩토링했습니다.

트랜잭션 처리는 두 도메인이 함께 변경되어야 하는 시나리오(**구독 요청 시 스트리머가 없으면 생성 후 구독**)를 `UseCase` 내부에서 묶어줌으로써, 한 도메인의 로직이 타 도메인을 침범하지 않도록 책임을 분리했습니다. 🙂

smileboy0014 reviewed (commented):
리뷰 남겼습니다!!

leesm0218 commented on src/infrastructure/streamer/client/streaming-server.core-client.ts:
```diff
@@ -149,4 +156,76 @@ export class StreamingServerCoreClient implements StreamingServerClient {
+        throw new RequestCustomException(RequestErrorCode.INVALID_CHANNEL_URL);
+      }
+
+      const chzzkBaseUrl = process.env.CHZZK_BASE_URL;
```
혹시 다른곳에서 환경변수 검사를 따로 하고계신가요?

soo96 replied:
```ts
const CHZZK_BASE_URL = process.env.CHZZK_BASE_URL!;
```

이런식으로 단언연산자(!) 처리 했습니다.
감사합니다🙏🏻

leesm0218 reviewed (approved):
수고하셨습니다.

soo96 committed:
refactor: PR 리뷰 반영

- repository, service 조회 함수 네이밍 get으로 통일
- 구독로직 Service -> UseCase로 이동
- SubscribeUseCase 구독 로직에 트랜잭션 적용

soo96 committed:
chore: PR 리뷰 반영 - 환경변수에 null 단언 연산자 적용

smileboy0014 reviewed (approved):
고생하셨습니다!!!
